### PR TITLE
Add soft debugger support to s390x

### PR DESCRIPTION
**Pull request by nealef**
* Add soft debugger support to s390x
  - Implement the sdb trampoline
  - Big-endian fixes for ppdb debugger:
    - Fix utf16 strings
    - Fix decoding of ImageDebugDirectory structures
  - New implementation of OP_SEQ_POINT
  - Enable successful unwind by defining markpoints
  - Implement mono_arch_setup_resume_sighandler_ctx
  - Implement unwind ops in epilog

* Fix delegate-invoke processing
  - Handle up to 10 parameters
  - Use sigparam information to correctly "slide" parameters

* Fix floating point negative zero for s390x
  - Add a new instruction for use in OP_RxCONST -0
  - Check for -0 in constant

* Minor indentation fixes

* Fix atomic add implementation

* Check for additional hardware facilities

### CODE REVIEW

File: mono/mini/unwind.c [hunk 1]
```diff
@@ -310,6 +310,7 @@ mono_print_unwind_info (guint8 *unwind_info, int unwind_info_len)
 		switch (op) {
 		case DW_CFA_advance_loc:
 			pos += *p & 0x3f;
+			printf ("CFA: [%x] advance loc\n",pos);
 			p ++;
```
Comment (vargaz) [inline mono/mini/unwind.c:314]: Its ok, but we don't print it because every instruction is prefixed by it.

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -4972,7 +4987,11 @@ debugger_agent_breakpoint_from_context (MonoContext *ctx)
 		return;
 
 	orig_ip = (guint8 *)MONO_CONTEXT_GET_IP (ctx);
+#ifndef __s390x__
 	MONO_CONTEXT_SET_IP (ctx, orig_ip - 1);
+#else
```
Comment (vargaz) [inline mono/mini/debugger-agent.c:4992]: Why is this needed ?

**Comment by vargaz** [comment]
Looks ok otherwise.

File: mono/mini/debugger-agent.c (hunk 2 above)
Comment (nealef) [reply mono/mini/debugger-agent.c:4992]: S390x call instructions are 2, 4, or 5 bytes in length. The one used for these calls are 2 bytes.

File: mono/mini/unwind.c (hunk 1 above)
Comment (nealef) [reply mono/mini/unwind.c:314]: I added it so I could do a 1 for 1 comparison with what I was specifying and what was being processed. It took a read of the code to see why I wasn't seeing this so I just added it for completeness (least astonishment principle).

File: mono/mini/debugger-agent.c (hunk 2 above)
Comment (vargaz) [reply mono/mini/debugger-agent.c:4992]: This is supposed to make sure that ip is in the previous instruction, it shouldn't depend on the instruction size, many other arches have 2/4 byte instructions.

Comment (nealef) [reply mono/mini/debugger-agent.c:4992]: Are you saying it's not dependent on it being set to the starting address of the instruction just that it's enough to be somewhere in that instruction?

Comment (vargaz) [reply mono/mini/debugger-agent.c:4992]: Yes.

Comment (vargaz) [reply mono/mini/debugger-agent.c:4992]: If that doesn't work, there might be an issue somewhere else.

Comment (nealef) [reply mono/mini/debugger-agent.c:4992]: I'll build it without that change and see how it goes. If it works for the other archs then I don't see why it won't work here. I just assumed it needed to point at the start of the instruction. What does it use that value for elsewhere in the agent?

Comment (vargaz) [reply mono/mini/debugger-agent.c:4992]: Mainly to compute which sequence point we are stopped at i think.

Comment (nealef) [reply mono/mini/debugger-agent.c]: Pushed change to revert the special case. Tests still pass (bar the same 1 that failed before).
